app/src/main/java/chat/rocket/android/chatroom/presentation/ChatRoomPresenter.kt

```diff
@@ -468,49 +467,82 @@ class ChatRoomPresenter @Inject constructor(
 
     fun uploadImage(roomId: String, mimeType: String, uri: Uri, bitmap: Bitmap, msg: String) {
         launchUI(strategy) {
-            view.showLoading()
+            val fileName = uriInteractor.getFileName(uri) ?: uri.toString()
             try {
                 withContext(Dispatchers.Default) {
-                    val fileName = uriInteractor.getFileName(uri) ?: uri.toString()
                     if (fileName.isEmpty()) {
                         view.showInvalidFileMessage()
                     } else {
                         val byteArray =
                             bitmap.getByteArray(mimeType, 60, settings.uploadMaxFileSize())
-                        retryIO("uploadFile($roomId, $fileName, $mimeType") {
-                            client.uploadFile(
-                                roomId,
-                                fileName,
-                                mimeType,
-                                msg,
-                                description = fileName
-                            ) {
-                                byteArray.inputStream()
-                            }
-                        }
 
+                        val id = UUID.randomUUID().toString()
+                        val username = userHelper.username()
+                        val user = userHelper.user()
+
+                        val attachment = Attachment(title=fileName, type="file", description = fileName, titleLink=uri.toString(), titleLinkDownload = true, imageUrl = uri.toString())
+                        val message = Message(
+                                id = id,
+                                attachments = listOf(attachment),
+                                roomId=roomId,
+                                timestamp = Instant.now().toEpochMilli(),
+                                sender = SimpleUser(user?.id, user?.username ?: username, user?.name),
+                                avatar = currentServer.avatarUrl(username ?: ""),
+                                synced = false,
+                                unread = true
+                        )
+                        messagesRepository.save(message)
+                        view.showNewMessage(
+                                mapper.map(
+                                        message,
+                                        RoomUiModel(roles = chatRoles, isBroadcast = chatIsBroadcast)
+                                ), false
+                        )
+
+                        client.uploadFile(
+                                roomId=roomId,
+                                fileName = fileName,
+                                mimeType = mimeType,
+                                msg = msg,
+                                description = fileName,
+                                id = id
```

Comment:
Why did you need to add 'id' to the upload file method?

Reply:
We generate id of the message on the client side, save a temp message object, and render it. We pass the id as an arg to create the message on the server side with the same id, then when we receive the message from the server, we check if a message with same id is already rendered or not, if yes, replace it with the one received from the server.